java-dynamic-sqs-listener-core/src/main/java/com/jashmore/sqs/retriever/prefetch/PrefetchingMessageRetriever.java

```diff
@@ -0,0 +1,216 @@
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.TimeUnit;
+import javax.validation.constraints.NotNull;
+
+/**
+ * Message retriever that allows for the pre-fetching of messages for faster throughput by making sure that there are always messages in a queue locally to be
+ * pulled from when one is needed.
+ *
+ * <p>The way this works is via the usage of an internal {@link BlockingDeque} that is used to store the prefetched of messages. When the limit of messages to
+ * prefetch has been hit the pushing of the message onto the queue will be blocked until a consumer is able to consume it. This means that the internal
+ * implementation is a bit odd and could be hard to debug when something is going wrong.
+ *
+ * <p>For the explanation, a {@link PrefetchingMessageRetriever} is built with a min prefetch size of 8, maximum prefetch size of 15 and a max of 10 messages
+ * able to be obtained in one call to SQS. The events that occur for this retriever are:
+ *
+ * <ol>
+ *     <li>{@link PrefetchingMessageRetriever} constructor is called</li>
+ *     <li>To accommodate this the internal queue is only made to be of size 7, which is one less than the min number of prefetched messages.  This is
+ *         important.</li>
+ *     <li>The {@link #start()} method is called which starts the prefetching of messages from SQS process.</li>
+ *     <li>A request is made out to retrieve 10 messages from SQS. This was defined by the
+ *         {@link PrefetchingProperties#getMaxNumberOfMessagesToObtainFromServer()} value passed into the constructor.</li>
+ *     <li>SQS responds with 10 messages.</li>
+ *     <li>7 messages are placed into {@link #internalMessageQueue} but no more able to be placed in due to the limit of the queue and therefore the
+ *         {@link QueueMessageRetriever} thread is blocked until messages are consumed before placing the other 3 messages onto the queue.</li>
+ *     <li>As the consumers begin to consume the messages from this retriever the queue begins to shrink in size until there are 7 messages in the queue but
+ *         zero needing to be placed in from the original request.</li>
+ *     <li>At this point the {@link QueueMessageRetriever} is free to go out and obtain more messages from SQS and it will attempt to retrieve
+ *         8 messages (this is due to the limit of 15 messages at maximum being prefetched)</li>
+ *     <li>This process repeats as more messages are consumed and placed onto the queues.</li>
+ * </ol>
+ *
+ * <p>Note that because these messages are being prefetched they could be in the internal queue for a long period and could even remain in the prefetched queue
+ * after the visibility timeout for the message has expired. This could cause it to be placed in the dead letter queue or attempted again at a future time.
+ */
+@Slf4j
+public class PrefetchingMessageRetriever implements AsyncMessageRetriever {
```

Comment:
Can't you get the same effect from just using the single fetcher w/ > 1 size of max messages to SQS request? You could crank up the size of the concurrent broker and that would have similar effect. 

Personally, I would leave this complication out for v1 and concentrate on the simple case and add this one in later.

Reply:
A ConcurrentMessageBroker with a IndividualMessageRetriever would not be work on this because each thread would be calling out to retrieve a single message from the queue.

As the IndividualMessageRetriever has no internal cache of messages that can store any excess messages retrieved from a larger than 1 size of max messages these extra messages would be lost. 

This retriever type aligns with how the JMS SQS listener currently works where it prefetches issues to be in the queue and this would theoretically be more performant as it is already trying to download more messages while some are being processed.

The middleground between the IndividualMessageRetriever and the PrefetchingMessageRetriever is the BatchingMessageRetriever which will group multiple requests for a message into a single call. See https://github.com/JaidenAshmore/java-dynamic-sqs-listener/issues/4  E.g. it waits a max of 1 second and once that time limit it reaches it will take all of the threads requesting messages and call out to SQS to obtain that total number of messages. 

One thing I wanted to provide in this implementation is feature parity with the existing frameworks so providing this Prefetching implementation is needed. But more importantly I just enjoyed working on it and want to keep it ;) I still want to figure out how to simplify it internally, maybe with Semaphores and a unbounded blocking queue instead of a bounded blocking queue.